File: python/src/chiptime/encode.py
"""FIT encoder — canonical wire form (ADR-0006).

Two producers feed `encode_messages`:
- `encodable_from_message`: lossless re-emit of a decoded Message (unknown
  content included; compressed-header timestamps materialized as field 253).
- `encodable_from_profile`: synthesize a message from profile names/values
  (repair's session/activity/events).

This is a programming surface, not a hostile-input surface: bad inputs raise
EncodeError instead of becoming defects.
"""

from __future__ import annotations

import re
import struct
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any

from chiptime.decode import FIT_EPOCH_UNIX
from chiptime.frames import crc16
from chiptime.message import Message
from chiptime.profile import BASE_TYPES, BASE_TYPES_BY_NAME, ENUMS, MESSAGES

HEADER_SIZE = 14
PROTOCOL_VERSION = 0x20
PROFILE_VERSION = 21141


class EncodeError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class FieldSpecValue:
    num: int
    base_type: int
    raw: Any  # int | float | bytes | list — wire-ready (sentinel substitution done)
    size: int


@dataclass(frozen=True, slots=True)
class DevSpecValue:
    num: int
    size: int
    dev_index: int
    raw: bytes


@dataclass(frozen=True, slots=True)
class EncodableMessage:
    global_num: int
    specs: tuple[FieldSpecValue, ...]
    dev_specs: tuple[DevSpecValue, ...] = ()

    @property
    def shape(self) -> tuple[Any, ...]:
        return (
            self.global_num,
            tuple((s.num, s.base_type, s.size) for s in self.specs),
            tuple((d.num, d.size, d.dev_index) for d in self.dev_specs),
        )


@dataclass
class _Slots:
    by_shape: dict[tuple[Any, ...], int] = field(default_factory=dict)
    shapes: dict[int, tuple[Any, ...]] = field(default_factory=dict)
    next_slot: int = 0

    def get(self, shape: tuple[Any, ...]) -> tuple[int, bool]:
        """Return (local_id, needs_definition)."""
        if shape in self.by_shape:
            return self.by_shape[shape], False
        local = self.next_slot
        self.next_slot = (self.next_slot + 1) % 16
        old = self.shapes.get(local)
        if old is not None:
            del self.by_shape[old]
        self.by_shape[shape] = local
        self.shapes[local] = shape
        return local, True


def encode_messages(messages: list[EncodableMessage]) -> bytes:
    body = bytearray()
    slots = _Slots()
    for em in messages:
        local, needs_def = slots.get(em.shape)
        if needs_def:
            body += _definition(local, em)
        body += _data(local, em)

    head = bytearray([HEADER_SIZE, PROTOCOL_VERSION])
    head += struct.pack("<H", PROFILE_VERSION)
    head += struct.pack("<I", len(body))
    head += b".FIT"
    head += struct.pack("<H", crc16(bytes(head[:12])))
    out = bytes(head) + bytes(body)
    return out + struct.pack("<H", crc16(out))


def _definition(local: int, em: EncodableMessage) -> bytes:
    if len(em.specs) > 255:
        raise EncodeError(f"message {em.global_num}: too many fields")
    hdr = 0x40 | (0x20 if em.dev_specs else 0x00) | local
    out = bytearray([hdr, 0, 0])  # little-endian always (ADR-0006)
    out += struct.pack("<H", em.global_num)
    out.append(len(em.specs))
    for s in em.specs:
        if not 0 < s.size < 256:
            raise EncodeError(f"field {s.num}: size {s.size} out of range")
        out += bytes([s.num, s.size, s.base_type])
    if em.dev_specs:
        out.append(len(em.dev_specs))
        for d in em.dev_specs:
            out += bytes([d.num, d.size, d.dev_index])
    return bytes(out)


def _data(local: int, em: EncodableMessage) -> bytes:
    out = bytearray([local])
    for s in em.specs:
        bt = BASE_TYPES.get(s.base_type)
        if bt is None or bt.struct_code is None:  # string/byte/unknown → raw bytes
            raw = s.raw if isinstance(s.raw, bytes) else bytes(s.raw)
            if len(raw) > s.size:
                raise EncodeError(f"field {s.num}: {len(raw)} bytes exceeds size {s.size}")
            out += (
                raw + b"\xff" * (s.size - len(raw))
                if bt is None
                else (raw + b"\x00" * (s.size - len(raw)))
            )
            continue
        count = s.size // bt.size
        vals = s.raw if isinstance(s.raw, (list, tuple)) else [s.raw]
        if len(vals) > count:
            raise EncodeError(f"field {s.num}: {len(vals)} values exceed count {count}")
        vals = list(vals) + [None] * (count - len(vals))
        for v in vals:
            if v is None:
                if bt.name in ("float32", "float64"):
                    # the exact invalid pattern, not an arbitrary NaN (muktihari#39)
                    out += b"\xff" * bt.size
                    continue
                v = _invalid_raw(bt.name)
            try:
                out += struct.pack("<" + bt.struct_code, v)
            except struct.error as exc:
                raise EncodeError(f"field {s.num}: {v!r} does not fit {bt.name}") from exc
    for d in em.dev_specs:
        if len(d.raw) != d.size:
            raise EncodeError(f"dev field {d.num}: {len(d.raw)} bytes != size {d.size}")
        out += d.raw
    return bytes(out)


def _invalid_raw(bt_name: str) -> int | float:
    from chiptime.profile.base_types import _SIGNED_INVALID
    from chiptime.profile.base_types import BASE_TYPES_BY_NAME as _BT

    if bt_name in ("float32", "float64"):
        return float("nan")  # canonical invalid pattern via NaN packing
    if bt_name in _SIGNED_INVALID:
        return _SIGNED_INVALID[bt_name]
    inv = _BT[bt_name].invalid
    assert inv is not None
    return inv


# ── producers ───────────────────────────────────────────────────────────────


def encodable_from_message(msg: Message) -> EncodableMessage:
    """Lossless re-emit from a decoded message's wire definition + raw values."""
    if msg.wire is None:
        raise EncodeError(f"message {msg.name} has no wire definition; use encodable_from_profile")
    by_num: dict[int, Any] = {}
    dev_by_key: dict[tuple[int, int], Any] = {}
    mdef = MESSAGES.get(msg.global_num)
    name_to_num = {f.name: n for n, f in mdef.fields.items()} if mdef else {}
    for fname, fv in msg.fields.items():
        if fv.developer is not None:
            key = (fv.developer.developer_data_index, fv.developer.field_definition_number)
            dev_by_key[key] = fv.raw
            continue
        if fname == "timestamp":
            by_num[253] = fv.raw
        elif fname in name_to_num:  # profile name wins (field_description has
            by_num[name_to_num[fname]] = fv.raw  # a real field NAMED field_definition_number)
        elif (m := re.fullmatch(r"field_(\d+)", fname)) is not None:
            by_num[int(m.group(1))] = fv.raw
    specs: list[FieldSpecValue] = []
    seen_253 = False
    for ws in msg.wire.fields:
        seen_253 = seen_253 or ws.num == 253
        specs.append(FieldSpecValue(ws.num, ws.base_type, by_num.get(ws.num), ws.size))
    if not seen_253 and "timestamp" in msg.fields:
        # compressed-header timestamp materialized (ADR-0006 §2)
        ts_raw = msg.fields["timestamp"].raw
        specs.append(FieldSpecValue(253, 0x86, ts_raw, 4))
    dev_specs = tuple(
        DevSpecValue(
            ds.num,
            ds.size,
            ds.dev_data_index,
            _dev_bytes(dev_by_key.get((ds.dev_data_index, ds.num)), ds.size),
        )
        for ds in msg.wire.dev_fields
    )
    return EncodableMessage(msg.global_num, tuple(specs), dev_specs)


def _dev_bytes(raw: Any, size: int) -> bytes:
    """Re-pack a decoded dev-field raw value into its wire bytes."""
    if raw is None:
        return b"\xff" * size
    if isinstance(raw, bytes):
        if len(raw) != size:
            raise EncodeError(f"dev field bytes {len(raw)} != size {size}")
        return raw
    if isinstance(raw, int):
        return raw.to_bytes(size, "little", signed=raw < 0)
    if isinstance(raw, float):
        code = {4: "<f", 8: "<d"}.get(size)
        if code is None:
            raise EncodeError(f"float dev field with size {size}")
        return struct.pack(code, raw)
    if isinstance(raw, (list, tuple)):
        if not raw:
            return b"\xff" * size
        per = size // len(raw)
        return b"".join(_dev_bytes(v, per) for v in raw)
    raise EncodeError(f"cannot re-pack dev value {raw!r}")


def encodable_from_profile(global_num: int, values: dict[str, Any]) -> EncodableMessage:
    """Synthesize a message from profile field names + semantic values."""
    mdef = MESSAGES.get(global_num)
    if mdef is None:
        raise EncodeError(
            f"unknown global message {global_num}; profile synthesis needs a known message"
        )
    by_name = {f.name: f for f in mdef.fields.values()}
    specs: list[FieldSpecValue] = []
    for fname, value in values.items():
        fdef = by_name.get(fname)
        if fdef is None:
            raise EncodeError(f"{mdef.name} has no field {fname!r}")
        bt_name, raw, size = _reverse(fdef, value)
        specs.append(FieldSpecValue(fdef.num, BASE_TYPES_BY_NAME[bt_name].byte, raw, size))
    return EncodableMessage(global_num, tuple(specs))


# Wire types for profile-synthesized fields (canonical choices).
_SYNTH_TYPES: dict[str, str] = {
    "timestamp": "uint32",
    "start_time": "uint32",
    "local_timestamp": "uint32",
    "time_created": "uint32",
}


def _reverse(fdef: Any, value: Any) -> tuple[str, Any, int]:
    kind = fdef.kind
    if kind in ("date_time", "local_date_time"):
        if isinstance(value, datetime):
            raw = int(value.timestamp()) - FIT_EPOCH_UNIX
        elif isinstance(value, int):
            raw = value
        else:
            raise EncodeError(f"{fdef.name}: expected datetime or FIT seconds")
        return "uint32", raw, 4
    if kind.startswith("enum:"):
        mapping = ENUMS.get(kind.removeprefix("enum:"), {})
        if isinstance(value, str):
            rev = {v: k for k, v in mapping.items()}
            if value not in rev:
                raise EncodeError(f"{fdef.name}: unknown enum name {value!r}")
            raw = rev[value]
        else:
            raw = int(value)
        ebt = "uint16" if fdef.name in ("manufacturer", "manufacturer_id") else "enum"
        return ebt, raw, BASE_TYPES_BY_NAME[ebt].size
    if kind == "string":
        sraw = str(value).encode("utf-8") + b"\x00"
        return "string", sraw, len(sraw)
    if kind == "bytes":
        braw = bytes(value)
        return "byte", braw, len(braw)
    # numbers: reverse scale/offset; choose a wide-enough canonical type
    raw_num = round((float(value) + fdef.offset) * fdef.scale)
    bt = _SYNTH_TYPES.get(fdef.name)
    if bt is None:
        if fdef.name in ("message_index", "num_laps", "first_lap_index"):
            bt = "uint16"
        elif raw_num < 0:
            bt = "sint32"
        elif raw_num <= 0xFE:
            bt = "uint8"
        elif raw_num <= 0xFFFE:
            bt = "uint16"
        else:
            bt = "uint32"
    return bt, raw_num, BASE_TYPES_BY_NAME[bt].size

File: js/src/encode.ts
/**
 * FIT encoder — canonical wire form (ADR-0006).
 *
 * Twin of `python/src/chiptime/encode.py`. Two producers feed `encodeMessages`:
 * - `encodableFromMessage`: lossless re-emit of a decoded Message (unknown content
 *   included; compressed-header timestamps materialized as field 253).
 * - `encodableFromProfile`: synthesize a message from profile names/values
 *   (repair's session/activity/events).
 *
 * This is a programming surface, not a hostile-input surface: bad inputs throw
 * `EncodeError` instead of becoming defects.
 */

import { crc16 } from "./frames.js";
import type { Message } from "./message.js";
import { pyRound } from "./numeric.js";
import { BASE_TYPES, BASE_TYPES_BY_NAME, ENUMS, MESSAGES } from "./profile/index.js";
import type { BaseType } from "./profile/index.js";

export const HEADER_SIZE = 14;
export const PROTOCOL_VERSION = 0x20;
export const PROFILE_VERSION = 21141;

export class EncodeError extends Error {
  override readonly name = "EncodeError";
}

/** Wire-ready field value (sentinel substitution done by the encoder). */
export interface FieldSpecValue {
  readonly num: number;
  readonly baseType: number;
  readonly raw: unknown; // number | bigint | Uint8Array | array | null
  readonly size: number;
}

export interface DevSpecValue {
  readonly num: number;
  readonly size: number;
  readonly devIndex: number;
  readonly raw: Uint8Array;
}

export interface EncodableMessage {
  readonly globalNum: number;
  readonly specs: readonly FieldSpecValue[];
  readonly devSpecs: readonly DevSpecValue[];
}

/** Python compares shape tuples; a string key is the JS equivalent. */
function shapeKey(em: EncodableMessage): string {
  const f = em.specs.map((s) => `${s.num},${s.baseType},${s.size}`).join(";");
  const d = em.devSpecs.map((s) => `${s.num},${s.size},${s.devIndex}`).join(";");
  return `${em.globalNum}|${f}|${d}`;
}

class Slots {
  private byShape = new Map<string, number>();
  private shapes = new Map<number, string>();
  private nextSlot = 0;

  /** Returns [localId, needsDefinition]. */
  get(shape: string): [number, boolean] {
    const existing = this.byShape.get(shape);
    if (existing !== undefined) return [existing, false];
    const local = this.nextSlot;
    this.nextSlot = (this.nextSlot + 1) % 16;
    const old = this.shapes.get(local);
    if (old !== undefined) this.byShape.delete(old);
    this.byShape.set(shape, local);
    this.shapes.set(local, shape);
    return [local, true];
  }
}

class ByteSink {
  private buf = new Uint8Array(1 << 12);
  len = 0;

  private ensure(extra: number): void {
    if (this.len + extra <= this.buf.length) return;
    let size = this.buf.length * 2;
    while (size < this.len + extra) size *= 2;
    const next = new Uint8Array(size);
    next.set(this.buf.subarray(0, this.len));
    this.buf = next;
  }

  byte(b: number): void {
    this.ensure(1);
    this.buf[this.len++] = b;
  }

  bytes(bs: Uint8Array | number[]): void {
    this.ensure(bs.length);
    for (const b of bs) this.buf[this.len++] = b as number;
  }

  u16le(v: number): void {
    this.ensure(2);
    this.buf[this.len++] = v & 0xff;
    this.buf[this.len++] = (v >> 8) & 0xff;
  }

  u32le(v: number): void {
    this.ensure(4);
    this.buf[this.len++] = v & 0xff;
    this.buf[this.len++] = (v >>> 8) & 0xff;
    this.buf[this.len++] = (v >>> 16) & 0xff;
    this.buf[this.len++] = (v >>> 24) & 0xff;
  }

  result(): Uint8Array {
    return this.buf.slice(0, this.len);
  }
}

/** Pack one numeric element of `bt` at little endian, Python `struct.pack` style. */
function packElement(out: ByteSink, bt: BaseType, v: number | bigint, fieldNum: number): void {
  const dv = new DataView(new ArrayBuffer(8));
  try {
    switch (bt.accessor) {
      case "getInt8":
        checkRange(v, -0x80, 0x7f, bt.name, fieldNum);
        dv.setInt8(0, Number(v));
        break;
      case "getUint8":
        checkRange(v, 0, 0xff, bt.name, fieldNum);
        dv.setUint8(0, Number(v));
        break;
      case "getInt16":
        checkRange(v, -0x8000, 0x7fff, bt.name, fieldNum);
        dv.setInt16(0, Number(v), true);
        break;
      case "getUint16":
        checkRange(v, 0, 0xffff, bt.name, fieldNum);
        dv.setUint16(0, Number(v), true);
        break;
      case "getInt32":
        checkRange(v, -0x80000000, 0x7fffffff, bt.name, fieldNum);
        dv.setInt32(0, Number(v), true);
        break;
      case "getUint32":
        checkRange(v, 0, 0xffffffff, bt.name, fieldNum);
        dv.setUint32(0, Number(v), true);
        break;
      case "getFloat32":
        dv.setFloat32(0, Number(v), true);
        break;
      case "getFloat64":
        dv.setFloat64(0, Number(v), true);
        break;
      case "getBigInt64":
        dv.setBigInt64(0, BigInt(v as number | bigint), true);
        break;
      case "getBigUint64":
        dv.setBigUint64(0, BigInt(v as number | bigint), true);
        break;
      default:
        throw new EncodeError(`field ${fieldNum}: cannot pack ${bt.name}`);
    }
  } catch (e) {
    if (e instanceof EncodeError) throw e;
    throw new EncodeError(`field ${fieldNum}: ${String(v)} does not fit ${bt.name}`);
  }
  for (let i = 0; i < bt.size; i++) out.byte(dv.getUint8(i));
}

function checkRange(v: number | bigint, lo: number, hi: number, name: string, num: number): void {
  // Python's struct raises on out-of-range and on non-integral ints alike.
  const n = typeof v === "bigint" ? v : v;
  if (typeof n === "number" && !Number.isInteger(n)) {
    throw new EncodeError(`field ${num}: ${n} does not fit ${name}`);
  }
  const bn = typeof n === "bigint" ? n : BigInt(Math.trunc(n as number));
  if (bn < BigInt(lo) || bn > BigInt(hi)) {
    throw new EncodeError(`field ${num}: ${n} does not fit ${name}`);
  }
}

const SIGNED_INVALID: Readonly<Record<string, number | bigint>> = {
  sint8: 0x7f,
  sint16: 0x7fff,
  sint32: 0x7fffffff,
  sint64: 0x7fffffffffffffffn,
};

function invalidRaw(btName: string): number | bigint {
  const signed = SIGNED_INVALID[btName];
  if (signed !== undefined) return signed;
  const inv = BASE_TYPES_BY_NAME[btName]?.invalid;
  if (inv === null || inv === undefined) {
    throw new EncodeError(`no invalid sentinel for ${btName}`);
  }
  return inv;
}

export function encodeMessages(messages: readonly EncodableMessage[]): Uint8Array {
  const body = new ByteSink();
  const slots = new Slots();
  for (const em of messages) {
    const [local, needsDef] = slots.get(shapeKey(em));
    if (needsDef) writeDefinition(body, local, em);
    writeData(body, local, em);
  }
  const bodyBytes = body.result();

  const head = new ByteSink();
  head.byte(HEADER_SIZE);
  head.byte(PROTOCOL_VERSION);
  head.u16le(PROFILE_VERSION);
  head.u32le(bodyBytes.length);
  head.bytes([0x2e, 0x46, 0x49, 0x54]); // ".FIT"
  const head12 = head.result();
  head.u16le(crc16(head12));

  const out = new ByteSink();
  out.bytes(head.result());
  out.bytes(bodyBytes);
  const whole = out.result();
  out.u16le(crc16(whole));
  return out.result();
}

function writeDefinition(out: ByteSink, local: number, em: EncodableMessage): void {
  if (em.specs.length > 255) {
    throw new EncodeError(`message ${em.globalNum}: too many fields`);
  }
  const hdr = 0x40 | (em.devSpecs.length > 0 ? 0x20 : 0x00) | local;
  out.bytes([hdr, 0, 0]); // little-endian always (ADR-0006)
  out.u16le(em.globalNum);
  out.byte(em.specs.length);
  for (const s of em.specs) {
    if (!(s.size > 0 && s.size < 256)) {
      throw new EncodeError(`field ${s.num}: size ${s.size} out of range`);
    }
    out.bytes([s.num, s.size, s.baseType]);
  }
  if (em.devSpecs.length > 0) {
    out.byte(em.devSpecs.length);
    for (const d of em.devSpecs) out.bytes([d.num, d.size, d.devIndex]);
  }
}

function writeData(out: ByteSink, local: number, em: EncodableMessage): void {
  out.byte(local);
  for (const s of em.specs) {
    const bt = BASE_TYPES[s.baseType];
    if (bt === undefined || bt.accessor === null) {
      // string/byte/unknown -> raw bytes
      const raw =
        s.raw instanceof Uint8Array
          ? s.raw
          : s.raw === null || s.raw === undefined
            ? new Uint8Array(0)
            : Uint8Array.from(s.raw as ArrayLike<number>);
      if (raw.length > s.size) {
        throw new EncodeError(`field ${s.num}: ${raw.length} bytes exceeds size ${s.size}`);
      }
      out.bytes(raw);
      const pad = bt === undefined ? 0xff : 0x00;
      for (let i = raw.length; i < s.size; i++) out.byte(pad);
      continue;
    }
    const count = Math.floor(s.size / bt.size);
    const vals: unknown[] = Array.isArray(s.raw) ? [...s.raw] : [s.raw];
    if (vals.length > count) {
      throw new EncodeError(`field ${s.num}: ${vals.length} values exceed count ${count}`);
    }
    while (vals.length < count) vals.push(null);
    for (const v of vals) {
      if (v === null || v === undefined) {
        if (bt.name === "float32" || bt.name === "float64") {
          // the exact invalid pattern, not an arbitrary NaN (muktihari#39)
          for (let i = 0; i < bt.size; i++) out.byte(0xff);
          continue;
        }
        packElement(out, bt, invalidRaw(bt.name), s.num);
        continue;
      }
      if (typeof v !== "number" && typeof v !== "bigint") {
        throw new EncodeError(`field ${s.num}: ${String(v)} does not fit ${bt.name}`);
      }
      packElement(out, bt, v, s.num);
    }
  }
  for (const d of em.devSpecs) {
    if (d.raw.length !== d.size) {
      throw new EncodeError(`dev field ${d.num}: ${d.raw.length} bytes != size ${d.size}`);
    }
    out.bytes(d.raw);
  }
}

// ── producers ───────────────────────────────────────────────────────────────

/** uint type code by width — the canonical numeric form for a reassembled field. */
const UINT_BY_SIZE: Readonly<Record<number, number>> = { 1: 0x02, 2: 0x84, 4: 0x86, 8: 0x8f };

/**
 * Base type to re-emit a field in.
 *
 * Normally the wire type, but decode sometimes *reassembles* a field a broken
 * encoder mis-declared — field 253 written as `byte[4]` when it is a timestamp
 * (taxonomy #17/#88, the Xiaomi-pipeline class). Emitting the numeric type the
 * value actually is produces the canonical wire form (ADR-0006); the
 * reinterpretation was already announced at parse time.
 */
function canonicalBaseType(baseType: number, raw: unknown, size: number): number {
  const bt = BASE_TYPES[baseType];
  if (bt !== undefined && bt.accessor !== null) return baseType; // already numeric
  const isInt = (typeof raw === "number" && Number.isInteger(raw)) || typeof raw === "bigint";
  if (!isInt) return baseType; // genuine strings/bytes are untouched
  return UINT_BY_SIZE[size] ?? baseType;
}

const FIELD_NUM_RE = /^field_(\d+)$/;

/** Lossless re-emit from a decoded message's wire definition + raw values. */
export function encodableFromMessage(msg: Message): EncodableMessage {
  if (msg.wire === null) {
    throw new EncodeError(`message ${msg.name} has no wire definition; use encodableFromProfile`);
  }
  const byNum = new Map<number, unknown>();
  const devByKey = new Map<string, unknown>();
  const mdef = MESSAGES[msg.globalNum];
  const nameToNum = new Map<string, number>();
  if (mdef) {
    for (const [n, f] of Object.entries(mdef.fields)) nameToNum.set(f.name, Number(n));
  }
  for (const [fname, fv] of msg.fields) {
    if (fv.developer !== null) {
      devByKey.set(
        `${fv.developer.developerDataIndex}:${fv.developer.fieldDefinitionNumber}`,
        fv.raw,
      );
      continue;
    }
    if (fname === "timestamp") {
      byNum.set(253, fv.raw);
    } else if (nameToNum.has(fname)) {
      // profile name wins (field_description has a real field NAMED field_definition_number)
      byNum.set(nameToNum.get(fname) as number, fv.raw);
    } else {
      const m = FIELD_NUM_RE.exec(fname);
      if (m !== null) byNum.set(Number(m[1]), fv.raw);
    }
  }
  const specs: FieldSpecValue[] = [];
  let seen253 = false;
  for (const ws of msg.wire.fields) {
    seen253 = seen253 || ws.num === 253;
    const raw = byNum.get(ws.num) ?? null;
    specs.push({
      num: ws.num,
      baseType: canonicalBaseType(ws.baseType, raw, ws.size),
      raw,
      size: ws.size,
    });
  }
  if (!seen253 && msg.fields.has("timestamp")) {
    // compressed-header timestamp materialized (ADR-0006 §2)
    const tsRaw = msg.fields.get("timestamp")?.raw ?? null;
    specs.push({ num: 253, baseType: 0x86, raw: tsRaw, size: 4 });
  }
  const devSpecs = msg.wire.devFields.map((ds) => ({
    num: ds.num,
    size: ds.size,
    devIndex: ds.devDataIndex,
    raw: devBytes(devByKey.get(`${ds.devDataIndex}:${ds.num}`) ?? null, ds.size),
  }));
  return { globalNum: msg.globalNum, specs, devSpecs };
}

/** Re-pack a decoded dev-field raw value into its wire bytes. */
function devBytes(raw: unknown, size: number): Uint8Array {
  if (raw === null || raw === undefined) return new Uint8Array(size).fill(0xff);
  if (raw instanceof Uint8Array) {
    if (raw.length !== size) {
      throw new EncodeError(`dev field bytes ${raw.length} != size ${size}`);
    }
    return raw;
  }
  if (typeof raw === "bigint" || (typeof raw === "number" && Number.isInteger(raw))) {
    // Python: raw.to_bytes(size, "little", signed=raw < 0)
    let v = typeof raw === "bigint" ? raw : BigInt(raw);
    const negative = v < 0n;
    if (negative) v += 1n << BigInt(size * 8); // two's complement
    const out = new Uint8Array(size);
    for (let i = 0; i < size; i++) {
      out[i] = Number(v & 0xffn);
      v >>= 8n;
    }
    if (v !== 0n) throw new EncodeError(`dev field value does not fit ${size} byte(s)`);
    return out;
  }
  if (typeof raw === "number") {
    const dv = new DataView(new ArrayBuffer(size));
    if (size === 4) dv.setFloat32(0, raw, true);
    else if (size === 8) dv.setFloat64(0, raw, true);
    else throw new EncodeError(`float dev field with size ${size}`);
    return new Uint8Array(dv.buffer);
  }
  if (Array.isArray(raw)) {
    if (raw.length === 0) return new Uint8Array(size).fill(0xff);
    const per = Math.floor(size / raw.length);
    const parts = raw.map((v) => devBytes(v, per));
    const out = new Uint8Array(size);
    let at = 0;
    for (const p of parts) {
      out.set(p, at);
      at += p.length;
    }
    return out;
  }
  throw new EncodeError(`cannot re-pack dev value ${String(raw)}`);
}

/** Wire types for profile-synthesized fields (canonical choices). */
const SYNTH_TYPES: Readonly<Record<string, string>> = {
  timestamp: "uint32",
  start_time: "uint32",
  local_timestamp: "uint32",
  time_created: "uint32",
};

/** Synthesize a message from profile field names + semantic values. */
export function encodableFromProfile(
  globalNum: number,
  values: Readonly<Record<string, unknown>>,
): EncodableMessage {
  const mdef = MESSAGES[globalNum];
  if (mdef === undefined) {
    throw new EncodeError(
      `unknown global message ${globalNum}; profile synthesis needs a known message`,
    );
  }
  const byName = new Map<
    string,
    { num: number; name: string; kind: string; scale: number; offset: number }
  >();
  for (const f of Object.values(mdef.fields)) byName.set(f.name, f);
  const specs: FieldSpecValue[] = [];
  for (const [fname, value] of Object.entries(values)) {
    const fdef = byName.get(fname);
    if (fdef === undefined) {
      throw new EncodeError(`${mdef.name} has no field '${fname}'`);
    }
    const [btName, raw, size] = reverse(fdef, value);
    const bt = BASE_TYPES_BY_NAME[btName] as BaseType;
    specs.push({ num: fdef.num, baseType: bt.byte, raw, size });
  }
  return { globalNum, specs, devSpecs: [] };
}

const UTF8_ENCODER = new TextEncoder();

function reverse(
  fdef: { name: string; kind: string; scale: number; offset: number },
  value: unknown,
): [string, unknown, number] {
  const kind = fdef.kind;
  if (kind === "date_time" || kind === "local_date_time") {
    if (typeof value !== "number" || !Number.isInteger(value)) {
      throw new EncodeError(`${fdef.name}: expected FIT seconds`);
    }
    return ["uint32", value, 4];
  }
  if (kind.startsWith("enum:")) {
    const mapping = ENUMS[kind.slice(5)] ?? {};
    let raw: number;
    if (typeof value === "string") {
      let found: number | null = null;
      for (const [k, v] of Object.entries(mapping)) {
        if (v === value) found = Number(k); // last match wins, as Python's dict-reverse does
      }
      if (found === null) {
        throw new EncodeError(`${fdef.name}: unknown enum name '${value}'`);
      }
      raw = found;
    } else {
      raw = Math.trunc(Number(value));
    }
    const ebt = fdef.name === "manufacturer" || fdef.name === "manufacturer_id" ? "uint16" : "enum";
    return [ebt, raw, (BASE_TYPES_BY_NAME[ebt] as BaseType).size];
  }
  if (kind === "string") {
    const bytes = UTF8_ENCODER.encode(String(value));
    const withNul = new Uint8Array(bytes.length + 1);
    withNul.set(bytes);
    return ["string", withNul, withNul.length];
  }
  if (kind === "bytes") {
    const braw = value instanceof Uint8Array ? value : Uint8Array.from(value as ArrayLike<number>);
    return ["byte", braw, braw.length];
  }
  // numbers: reverse scale/offset; choose a wide-enough canonical type.
  // pyRound: Python's round() is half-to-even and this reaches the wire.
  const rawNum = pyRound((Number(value) + fdef.offset) * fdef.scale);
  let bt = SYNTH_TYPES[fdef.name];
  if (bt === undefined) {
    if (
      fdef.name === "message_index" ||
      fdef.name === "num_laps" ||
      fdef.name === "first_lap_index"
    ) {
      bt = "uint16";
    } else if (rawNum < 0) {
      bt = "sint32";
    } else if (rawNum <= 0xfe) {
      bt = "uint8";
    } else if (rawNum <= 0xfffe) {
      bt = "uint16";
    } else {
      bt = "uint32";
    }
  }
  return [bt, rawNum, (BASE_TYPES_BY_NAME[bt] as BaseType).size];
}
